Mathlib/Lean/Meta/RefinedDiscrTree/Basic.lean

```diff
@@ -1,178 +1,266 @@
+  cache    : AssocList Expr (List Key) := {}
+
+instance : Inhabited (LazyEntry) where
+  default := { mctx := {} }
 
+private def LazyEntry.format (entry : LazyEntry) : Format :=
+  let results := if entry.results matches [] then f!"" else f!", results: {entry.results}"
+  let todo    := if let some info := entry.previous then f!", todo: {info.expr}" else ""
+  f!"stack: {entry.stack}{results}{todo}"
+
+instance : ToFormat LazyEntry := ⟨LazyEntry.format⟩
+
+/-- Array index of a `Trie α` in the `tries` of a `RefinedDiscrTree`. -/
+abbrev TrieIndex := Nat
+
+/--
+Discrimination tree trie. See `RefinedDiscrTree`.
+
+A `Trie` will normally have exactly one of the following
+- nonempty `values`
+- nonempty`stars` or `children`
+- nonempty `pending`
+But defining it as a structure that can have all at the same time turns out to be easier.
+-/
+structure Trie (α : Type) where
+  node ::
+    /-- Return values, at a leaf -/
+    values : Array α
+    /-- Following `Trie`s based on a `Key.star`. -/
+    stars : Std.HashMap Nat TrieIndex
+    /-- Following `Trie`s based on the `Key`. -/
+    children : Std.HashMap Key TrieIndex
+    /-- Lazy entries that still have to be evaluated. -/
+    pending : Array (LazyEntry × α)
+
+instance {α : Type} : Inhabited (Trie α) := ⟨.node #[] {} {} #[]⟩
+
+end RefinedDiscrTree
+
+open RefinedDiscrTree in
+
+/--
+Discrimination tree. It is an index from expressions to values of type `α`.
+
+We store all of the nodes in one `Array`, `tries`, instead of using a 'normal' inductive type.
+This is so that we can modify the tree globally, which is very useful when evaluating lazy
+entries and saving the result globally.
+-/
+structure RefinedDiscrTree (α : Type) where
+  /-- `Trie`s at the root based of the `Key`. -/
+  root : Std.HashMap Key TrieIndex := {}
+  /-- Array of trie entries. Should be owned by this trie. -/
+  tries : Array (Trie α) := #[]
+
+namespace RefinedDiscrTree
+
+variable {α : Type}
 
-/-- Discrimination tree. It is an index from expressions to values of type `α`. -/
-structure _root_.Lean.Meta.RefinedDiscrTree (α : Type) where
-  /-- The underlying `PersistentHashMap` of a `RefinedDiscrTree`. -/
-  root : PersistentHashMap Key (Trie α) := {}
 instance : Inhabited (RefinedDiscrTree α) := ⟨{}⟩
 
-private partial def format [ToFormat α] (d : RefinedDiscrTree α) : Format :=
-  let (_, r) := d.root.foldl
-    (fun (p : Bool × Format) k c =>
-      (false,
-        p.2 ++ (if p.1 then Format.nil else Format.line) ++
-          Format.paren (Std.format k ++ " => " ++ Std.format c)))
-    (true, Format.nil)
-  Format.group r
+private partial def format [ToFormat α] (tree : RefinedDiscrTree α) : Format :=
+  let lines := tree.root.fold (init := #[]) fun lines key trie =>
+    lines.push (Format.nest 2 f! "{key} =>{Format.line}{go trie}")
+  if lines.size = 0 then
+    f! "<empty discrimination tree>"
+  else
+    lines.foldl (init := "Discrimination tree flowchart:") (· ++ Format.line ++ ·)
```

Comment:
I think this is
```suggestion
    Format.line.joinsep (f!"Discrimination tree flowchart:" :: lines)
```
note also that `Format.line` does not actually *require* a line break; if you want to prevent this appearing on one line if `lines` is short, use `\n`.